### ISSUE TITLE
RFC: In-memory sparse array interchange

### ISSUE THREAD

**Issue by hameerabbasi** [data-apis/array-api#840]
## Motivation
The [`sparse`](https://github.com/pydata/sparse) (also called PyData/Sparse) development team have been working on integration efforts with the ecosystem, most notably with SciPy, scikit-learn and others, with CuPy, PyTorch, JAX and TensorFlow also on the radar. One of the challenges we were facing was the lack of (possibly zero-copy) interchange between the different sparse array implementations. We believe this may be a pain point for many sparse array implementations moving forward.

This mirrors an issue seen for dense arrays previously, where the [DLPack protocol](https://dmlc.github.io/dlpack/latest/python_spec.html) was the one of the first things to be standardised. We're hoping to achieve community consensus for a similar problem.

Luckily, all sparse array formats (with the possible exception of DOK) are usually collections of dense arrays underneath. In addition, this problem has been solved for on-disk arrays before by the [binsparse specification](https://graphblas.org/binsparse-specification/). @willow-ahrens is a co-author of that spec, and is also a collaborator for the `sparse` work.

## Proposal
We propose introducing two new methods to the array-API compliant sparse array objects (such as those in `sparse`), which are described below.

### `__binsparse_descriptor__`
Returns a `dict` equivalent to a parsed JSON [`binsparse` descriptor](https://graphblas.org/binsparse-specification/#descriptor) of an array.

### `__binsparse_dlpack__`
Returns a `dict` with keys equivalent to those in the [`binsparse` specification](https://graphblas.org/binsparse-specification), and values that are `__dlpack__` compatible arrays.

### Modifications to `from_dlpack`
If a library supports sparse arrays as well as dense arrays, its `from_dlpack` method should support accepting (when possible, zero-copy) versions of objects that follow this `__binsparse_dlpack__` protocol, and have an equivalent sparse format within the library.

### Psuedocode implementation
Here's a psuedocode example using two libraries, `xp1` and `xp2`, both supporting sparse arrays:

```python
# In library code:
xp2_sparray = xp2.from_dlpack(xp1_sparray, ...)

# This psuedocode impl is common between `xp1` and `xp2`
def from_dlpack(x: object, /, *, device: device | None = None, copy: bool | None = None) -> array:
    dlpack_impl = getattr(x, "__dlpack__", None)
    if dlpack_impl is not None:
        # Process dense case
        return arr
    
    binsparse_descriptor_impl = getattr(x, "__binsparse_descriptor__", None)
    if binsparse_descriptor_impl is None:
        raise TypeError(...)
    
    binsparse_descriptor = binsparse_descriptor_impl()
    sparse_type = _type_from_binsparse_descriptor(binsparse_descriptor)
    
    binsparse_dlpack_impl = getattr(x, "__binsparse_dlpack__", None)
    if binsparse_dlpack_impl is None:
        raise TypeError(...)
    
    binsparse_arrays: dict[str, SupportsDLPack] = binsparse_dlpack_impl()
    # Some renaming may need to happen somewhere to match internal naming of constituent arrays
    my_constituent_arrays = {
        key: from_dlpack(arr, device=device, copy=copy) for k, arr in binsparse_arrays.items()
    }
    return sparse_type.from_dense_arrays(**my_constituent_arrays)
```

Parallel implementation in `sparse`: https://github.com/pydata/sparse/pull/764

Pinging possibly interested parties:
* @mtsokol @ivirshup (for `scipy.sparse`)
* @willow-ahrens @mtsokol (from `binsparse` and `finch-tensor`/`sparse`)
* @leo-fang (for `cupyx.sparse`)
* @pearu (for `torch.sparse`)
* @jakevdp (for JAX/TensorFlow)

**Comment by brunoR1000** [comment]
This might help you
https://bit.ly/3yXXzbX
Passcode: d5kp14h
You must have gcc compiler installed

**Comment by pearu** [comment]
A couple of quick notes.

First, dlpack and binsparse are self-contained specifications where dlpack provides a protocol for sharing strided arrays between different array libraries and binsparse provides a unified description of various sparse formats. This proposal tries to glue these together by implicitly extending [DLPack protocol](https://dmlc.github.io/dlpack/latest/python_spec.html#syntax-for-data-interchange-with-dlpack) but that contradicts with the semantics of DLPack that ["describes the memory layout of dense, strided, n-dimensional arrays"](https://dmlc.github.io/dlpack/latest/python_spec.html#semantics). In addition, the Python Array API standard specifies that [DLPack should not support sparse arrays](https://data-apis.org/array-api/latest/design_topics/data_interchange.html#non-supported-use-cases).

My suggestion is to cook up a sparse array interchange protocol that may use dlpack and binsparse (as these are obviously relevant pieces of this problem) but in a cleaner way.

For instance, binsparse specifies that the binsparse-compatible object provides a 2-tuple `(<sparse array descriptor>, <list of sparse array data as strided arrays>)`. So, the appropriate place to deploy the dlpack protocol is to require that the arrays in the `<list of sparse array data as strided arrays>` must support the dlpack protocol.

For instance 2, introduce "Python specification of binsparse" (similar to [one in dlpack](https://dmlc.github.io/dlpack/latest/python_spec.html#syntax-for-data-interchange-with-dlpack)) that consists of
1. a `from_binsparse` function that accepts objects that implement the binsparse protocol per 2. below
2. a `__binsparse__` method on the objects representing sparse arrays that will return a 2-tuple `(<sparse array descriptor>, <list of sparse array data as strided arrays>)` which is used by the `from_binsparse` function of the consumer library.

Also, the specification should include the C-side interface as well.

What do you think?

**Comment by hameerabbasi** [comment]
While I'm on board with the proposed `__binsparse__` protocol, I'm of the opinion that sparse arrays should have as little user-facing API differences as possible as compared to strided arrays, i.e. using sparse arrays should be as frictionless from the Array API side as using strided arrays. Allocating a different function based on sparse and dense arrays (`from_dlpack` vs `from_binsparse`) seems counter to that.

I'm not opposed to a C-side specification, I'd welcome that so that the benefits can extend beyond the Python ecosystem. We do have to be mindful that the binsparse specification requires key-value maps, and we would need to maintain their relative order. Perhaps an iterable of 2-tuples makes more sense in this case.

**Comment by pearu** [comment]
> While I'm on board with the proposed `__binsparse__` protocol, I'm of the opinion that sparse arrays should have as little user-facing API differences as possible as compared to strided arrays, i.e. using sparse arrays should be as frictionless from the Array API side as using strided arrays.

Absolutely. I presume that users of sparse arrays are aware of advantages and disadvantages of using a particular sparse format over other sparse formats or strided arrays. Using a particular sparse format is a choice of optimization method that must be supported by user-facing API.

> Allocating a different function based on sparse and dense arrays (`from_dlpack` vs `from_binsparse`) seems counter to that.

`from_dlpack` is specifically defined for strided arrays, both by the Python Array API and dlpack. So, I find abusing the `from_dlpack` function for sparse arrays more confusing than providing a new API function which would follow the same design pattern that dlpack introduced: the pair (`from_foo`, `__foo__`) provides a minimal API for sharing data between provider and consumer libraries that support a protocol `foo`.

Btw, if one considers strided and sparse arrays semantically equivalent (as in PyTorch, for instance), a more intuitive approach would be to use `asarray` instead of `from_dlpack` as a common API function for constructing strided or sparse arrays from provider library objects. Although, Python Array API does not support that and existing implementations (PyTorch, JAX, ...) do not mix the functionalities of `asarray` and `from_dlpack`. I don't suggest to do that here either.

**Comment by hameerabbasi** [comment]
> Btw, if one considers strided and sparse arrays semantically equivalent (as in PyTorch, for instance), a more intuitive approach would be to use `asarray` instead of `from_dlpack` as a common API function for constructing strided or sparse arrays from provider library objects.

While I agree with a common API; the disadvantage I find in this approach is that since `from_dlpack` requires the producer to pass ownership to the consumer; `asarray` requires one to maintain the ownership of the consumed array. This, in turn, means that there's no way to have a zero-copy transfer using `asarray` using the `from_dlpack` function; i.e. one would have to do something like `def asarray(x, ...): return from_dlpack(copy(x), ...)` somewhere to maintain ownership semantics, incurring a copy.

My intention was to have an API that essentially could support zero-copy interchange across libraries, regardless of whether the consumed array was strided, sparse or something else.

**Comment by pearu** [comment]
> While I agree with a common API; the disadvantage I find in this approach is that since `from_dlpack` requires the producer to pass ownership to the consumer;

This is incorrect. When using `from_dlpack(x)`, the producer keeps owning the memory of `x`. See https://dmlc.github.io/dlpack/latest/python_spec.html#semantics.

**Comment by hameerabbasi** [comment]
> > While I agree with a common API; the disadvantage I find in this approach is that since `from_dlpack` requires the producer to pass ownership to the consumer;
> 
> This is incorrect. When using `from_dlpack(x)`, the producer keeps owning the memory of `x`. See https://dmlc.github.io/dlpack/latest/python_spec.html#semantics.

Ah, in that case; yes, `asarray` can use `from_dlpack` and a possible `from_binsparse`, and still be zero-copy. Thanks for the correction; `asarray` seems like the right format-agnostic API for this.

**Comment by hameerabbasi** [comment]
I've updated the issue with @pearu's feedback.

**Comment by willow-ahrens** [comment]
I'd like to additionally CC the binsparse contributors so they're aware of this effort.
@BenBrock
@eriknw
@ivirshup
@jim22k
@DrTimothyAldenDavis

**Comment by hameerabbasi** [comment]
There are other on-disk/text formats: [1](https://math.nist.gov/MatrixMarket/formats.html#MMformat) [2](https://math.nist.gov/MatrixMarket/formats.html#hb)

Both of these are not designed for in-memory interchange of data -- and have no support for multiple formats. I'll add these to the issue.

**Comment by willow-ahrens** [comment]
There's also the FROSTT format, which stores tensors in a minimal text encoding and suffers the same limitations of the above.

http://frostt.io/

**Comment by rgommers** [comment]
The relevant issue on the `binsparse` repo for an in-memory representation is [GraphBLAS/binsparse-specification#4](https://github.com/GraphBLAS/binsparse-specification/issues/4). That also mentions that Arrow has support for 2-D CSR/CSC/CSF/COO (no n-D, no chunking).

There's of course a bit of a chicken-and-egg problem here: it's hard to standardize anything that doesn't exist yet, but library implementors want to move ahead and implement something that is likely to be standardized later to avoid fragmentation or bc-breaking changes. 

So here is what I think what should happen:
- This issue should work out what the desired Python API will be for sparse in-memory interchange (the proposed analogue to DLPack with a dunder method like `__binsparse__` sounds about right; whether more is needed a la [`__dlpack_device__`](https://data-apis.org/array-api/latest/API_specification/generated/array_api.array.__dlpack_device__.html#dlpack-device) is TBD I think).
- The actual `binsparse` in-memory spec should be worked out on [GraphBLAS/binsparse-specification#4](https://github.com/GraphBLAS/binsparse-specification/issues/4).
    - Seems more appropriate to discuss there rather than here, since that's the project that aims to define this and has more history and I assume additional constraints that aren't relevant for this repo like aligment between the on-disk and in-memory metadata.
    - DLPack seems like a more appropriate choice that Arrow from the perspective of array libraries to me, due to broader support of relevant devices and dtypes, and being already in the array API standard and supported by all relevant array libraryes. I can imagine that `binsparse` could decide to support both though, in case there's a real-world need for Arrow support.
- Libraries should then go ahead and implement that.
- Once we have multiple implementations that seem to work well, we move the Python API worked out in this issue into the standard.

**Comment by pearu** [comment]
PyTorch have sparse formats that binsparse specifications does not specify as pre-defined formats. For instance, there are BSR/BSC (blocked CSR/CSC) formats, hybrid sparse formats (COO/CSR/CSC/BSR/BSC with values being strided tensors), sparse formats with batch dimensions (CSR/CSC/BSR/BSC with indices being multidimensional).

So, for PyTorch, the usefulness of the in-memory sparse array interchange format based on binsparse specification will depend on how easy is to define these blocked/hybrid/batch sparse formats using binsparse custom format support, or better yet, on adding blocked/hybrid/batch sparse formats as pre-defined to binsparse specification.

**Comment by pearu** [comment]
CC: @amjames (for torch.sparse)

**Comment by hameerabbasi** [comment]
I have two things to add here:

1. The GraphBLAS folks are now explicitly considering in-memory use-cases when designing the binsparse protocol; in particular the reshaping-transposition proposal is being considered, but it is as-yet unclear what form that will take.
2. I feel that `__binsparse__` and `__binsparse_format__` should be separate methods; the reasoning behind this is exactly the same as `__dlpack__` and `__dlpack_device__`:
  a. One might need to query the format before getting the constituent arrays out, as getting the arrays out could be an expensive operation for some cases.
  b.  There are many parallels in design for needing to query the format (device) independent of the actual arrays in `__dlpack__`, such as erroring out early and requesting a specific format.

Please let me know what folks think of 2, especially @pearu who proposed the original move to a 1-method design.

**Comment by DrTimothyAldenDavis** [comment]
This is an extremely important issue:

"One might need to query the format before getting the constituent arrays out, as getting the arrays out could be an expensive operation for some cases."

**Comment by pearu** [comment]
Recall, `__dlpack__` method returns a capsule that holds a structure that contains all the information about the array: shape, dtype, strides, pointer to storage, etc. This makes previous array interface protocols (`__array_interface__`, `__cuda_array_interface__`) that exposed the array info explicitly, unnecessary.

I'd consider `__binsparse_format__` a similar protocol to `__array_interface__` that exposes various information about the sparse format: shape information (sparse and dense dimensions), dtype, format name, and format specific data that supports dlpack protocol.

Assuming that `__binsparse__` protocol implementation will be similar to `__dlpack__` protocol implementations, that is, `__binsparse__` method returns a capsule that holds a structure that contains information about the sparse array. Hence, `__binsparse_format__` is unnecessary: the query about the sparse array format can be implemented as a query to the binsparse capsule object, similar to how https://github.com/pearu/pydlpack/tree/main implements a query to dlpack capsule, without the need to access the storage of data.

**Comment by hameerabbasi** [comment]
> the query about the sparse array format can be implemented as a query to the binsparse capsule object

Right, this assumes an O(1) conversion to a binsparse capsule (or equivalent) is always possible. However, this may not be the case: Libraries may support more formats than are supported by the binsparse protocol, and may need to perform an additional conversion.

**Comment by pearu** [comment]
> Libraries may support more formats than are supported by the binsparse protocol, and may need to perform an additional conversion.

I am not sure how `__binsparse_format__` will be helpful here. If a library implements a sparse format that binsparse protocol does not support, then the binsparse protocol will be useless anyway: binsparse protocol cannot implement conversion about a format that it does not support, any necessary conversion should be done by the provider library.
If you'd disagree, please provide more details about the `__binsparse_format__` specification as I may miss some points here.

**Comment by hameerabbasi** [comment]
> any necessary conversion should be done by the provider library

I think you've hit the nail on the head with this part -- a conversion might be necessary, which may make the cost of constructing a capsule O(n), and therefore we're left with two options if we want to guarantee an O(1) conversion (both require a `.asformat`):
* Raise on unsupported formats (which can take binsparse descriptors). The consumer must choose whether to call `.asformat` on error.
* Allow the consumer to query if they support the format via `__binsparse_format__` and then raise or call `.asformat` as necessary.

**Comment by pearu** [comment]
An exception with a helpful exception message is better than a silent expensive conversion, imho.

**Comment by rgommers** [comment]
The two-method format seems clearly preferred, indeed for the same reason as `__dlpack_device__`: allowing to introspect metadata for compatibility and "hand-shaking" (it allows the consumer to change its request when it supports multiple devices/formats/etc.).

@pearu the analogy with `__array_interface__` is not a good one. That method is numpy-specific, and the actual data is hence always already in memory and always in a single format (strided array on CPU). A `__binsparse__` call, just like a `__dlpack__` call, _requires_ the data to be in memory in a supported format, so for (for example) lazily evaluated formats or custom formats not part of the protocol, a call to `__binsparse__` may indeed require an expensive operation (and no, always raising an exception instead is definitely not the correct answer, there are always needs to do the expensive thing if no cheap thing is possible).

If anything, we've so far found a few corner cases where it would be helpful for more parts of the DLPack protocol to be introspectable separately from the actual "give me a capsule with a pointer to data in memory".

**Comment by pearu** [comment]
> A __binsparse__ call, just like a __dlpack__ call, requires the data to be in memory in a supported format

Sparse tensors of any format can be modeled as a pair
```
(shape-dtype-format specs, (<a list of dlpack-compatible objects that hold the indices data and values>))
```
(this is how we think about sparse tensors in PyTorch, for instance). Notice that `__binsparse__` structure *does not (need to)* hold raw pointers to memory which indeed is the case of the `__dlpack__` structure. Hence, `__binsparse__` structure is lazy by definition when using this mental model.

Lazy access to sparse tensors is about accessing `shape-dtype-format specs` without accessing the data of indices and values. Notice that the `__dlpack__` capsule that holds pointers to raw memory is constructed when the `__dlpack__` method is called (this is the point where data needs to be materialized in memory, not earlier). So, lazy access to sparse tensors shape-dtype-format data is possible if one will not trigger the creation of `__dlpack__` capsules (read: materialize data in memory) earlier than necessary.

In any case, the laziness property should be provided at the DLpack protocol level, not on binsparse protocol that just uses the dlpack protocol for accessing sparse tensor data.

**Comment by rgommers** [comment]
> Lazy access to sparse tensors is about accessing `shape-dtype-format specs` without accessing the data of indices and values.

This is the entire point of why two separate methods have to exist.  Your answer agrees that there should be this separation, you're just moving the separation between accessing metadata and data to a place it doesn't exist (`__dlpack__`). Coupling the introduction of a binsparse protocol to not-yet-proposed and probably infeasible changes to `__dlpack__` isn't a good idea. Separation of concerns is nice:) 

Please either go with the 2-method approach for binsparse itself instead, or propose a way of what a 1-method `__binsparse__` should return to keep things lazy without any changes to DLPack.

**Comment by hameerabbasi** [comment]
I've modified the issue description to use the two-method version.

**Comment by hameerabbasi** [comment]
Looping in @BenBrock.

Based on https://github.com/scipy/scipy/pull/22553#issuecomment-2674384445 We discussed in the bi-weekly Binsparse meeting yesterday where the spec should live -- It seems the most appropriate thing to do is (similar to DLPack) let the binsparse repo take care of the descriptor and the naming conventions of the arrays, and the array-api repo should specify how that descriptor is used in Python-land.

For context, all the binsparse spec requires from back-ends is:

1. A way to represent and interpret JSON (Python objects equivalent to the parsed JSON in our case).
2. A key-value store of 1D and 2D arrays (A `dict[str, SupportsDLPack]` in our case).

So this seems to be a good separation of concerns.

**Comment by rgommers** [comment]
That seems reasonable. I think a draft PR on this repo for review of the new methods in more detail would be useful. Based on feedback and working PRs to 2-3 array libraries that are all in sync, things can be tested and iterated on until both library and spec maintainers are happy, before merging anything.

**Comment by hameerabbasi** [comment]
PR is up at #912.

**Comment by hameerabbasi** [comment]
@rgommers Following the discussion of yesterday's meeting, one valid use-case for implementing ``__binsparse__`` on dense arrays might be for an array library like PyTorch or `finch-tensor`, which supports both dense and sparse arrays. In that case, it isn't clear that `asarray` should densify: Indeed, it should keep the format as-is and attempt to use `__dlpack__` to consume the data. However, in this case, `arr.__binsparse__(descriptor=some_format)` (or, to a user, `xp2.from_binsparse(xp1_arr, descriptor=some_format)`) might be suitable to convert to a given sparse array format, or even to a dense one.

Alternatively we can add `array.asformat(some_descriptor)` on both sparse and dense arrays if `hasattr(xp, 'sparse')` holds.

I'm open to both options.

**Comment by rgommers** [comment]
That rationale isn't logically consistent. `asarray(x_dense)` will always use `__dlpack__` and it's irrelevant whether `__binsparse__` is present.

Can you please start simple, with _only_ binsparse on sparse arrays, and then have a very concrete code example of where/how things can go wrong? 

> Alternatively we can add `array.asformat(some_descriptor)` on both sparse and dense arrays if `hasattr(xp, 'sparse')` holds.

As a reminder: we're only working on `__binsparse__` and `from_binsparse` to work out any potential kinks and guiding libraries with sparse arrays on _how_ to implement those two things in the same way. Once that is stable in multiple libraries and proven useful, we can add it to the standard. It's quite premature at this point to discuss a larger API surface.

`hasattr(xp, 'sparse')` is also a misconception. We need `hasattr(xp, 'from_binsparse')` or `hasattr(an_array, '__binsparse__')`.
